src/aics-image-viewer/components/ViewerStateProvider/index.tsx

```diff
@@ -0,0 +1,236 @@
+  // View mode: if we're switching to 2d, switch to volumetric rendering
+  viewMode: (prevSettings, viewMode) => {
+    const switchToVolumetric = viewMode !== ViewMode.threeD && prevSettings.renderMode === RenderMode.pathTrace;
+    return {
+      ...prevSettings,
+      viewMode,
+      renderMode: switchToVolumetric ? RenderMode.volumetric : prevSettings.renderMode,
+    };
+  },
+  // Render mode: if we're switching to pathtrace, turn off autorotate
+  renderMode: (prevSettings, renderMode) => ({
+    ...prevSettings,
+    renderMode,
+    autorotate: renderMode === RenderMode.pathTrace ? false : prevSettings.autorotate,
+  }),
+  // Autorotate: do not enable autorotate while in pathtrace mode
+  autorotate: (prevSettings, autorotate) => ({
+    ...prevSettings,
+    // The button should theoretically be unclickable while in pathtrace mode, but this provides extra security
+    autorotate: prevSettings.renderMode === RenderMode.pathTrace ? false : autorotate,
+  }),
+};
+
+const extractViewerSettings = (context: ViewerContextType): ViewerState => {
+  const {
+    channelSettings: _channelSettings,
+    changeViewerSetting: _changeViewerSetting,
+    changeChannelSetting: _changeChannelSetting,
+    changeMultipleChannelSettings: _changeMultipleChannelSettings,
+    applyColorPresets: _applyColorPresets,
+    ...settings
+  } = context;
+  return settings;
+};
+
+/** Changes a key in a given `ViewerState` object, keeping the object in a valid state and applying partial values */
+const applyChangeToViewerSettings = <K extends keyof ViewerState>(
+  viewerSettings: ViewerState,
+  key: K,
+  value: PartialIfObject<ViewerState[K]>
+): ViewerState => {
+  const changeHandler = VIEWER_SETTINGS_CHANGE_HANDLERS[key];
+
+  if (changeHandler) {
+    // This setting has a custom change handler. Let it handle creating a new state object.
+    return changeHandler(viewerSettings, value);
+  } else {
+    const setting = viewerSettings[key];
+    if (isObject(setting) && isObject(value)) {
+      // This setting is an object, and we may be updating it with a partial object.
+      return { ...viewerSettings, [key]: { ...setting, ...value } };
+    } else {
+      // This setting is regular. Update it the regular way.
+      return { ...viewerSettings, [key]: value };
+    }
+  }
+};
+
+const nullfn = (): void => {};
+
+const DEFAULT_VIEWER_CONTEXT: ViewerContextType = {
+  ...DEFAULT_VIEWER_SETTINGS,
+  channelSettings: [],
+  changeViewerSetting: nullfn,
+  setChannelSettings: nullfn,
+  changeChannelSetting: nullfn,
+  changeMultipleChannelSettings: nullfn,
+  applyColorPresets: nullfn,
+};
+
+const DEFAULT_VIEWER_CONTEXT_OUTER = { ref: { current: DEFAULT_VIEWER_CONTEXT } };
+
+type NoNull<T> = { [K in keyof T]: NonNullable<T[K]> };
+type ContextRefType = NoNull<React.MutableRefObject<ViewerContextType>>;
+
+export const ViewerStateContext = React.createContext<{ ref: ContextRefType }>(DEFAULT_VIEWER_CONTEXT_OUTER);
```

Comment:
I did this because, as I learned, refs always include `null` in their value type, since they're originally designed to pull DOM nodes out of the committed component and those nodes might not exist.